src/main/java/org/spongepowered/common/mixin/core/entity/player/MixinEntityPlayerMP.java

```diff
@@ -649,12 +651,24 @@ public void restorePacketItem(EnumHand hand) {
 
     @Override
     public Optional<Container> openInventory(Inventory inventory) throws IllegalArgumentException {
-        return Optional.ofNullable((Container) SpongeCommonEventFactory.displayContainer((EntityPlayerMP) (Object) this,
-                inventory));
+        if (((IMixinContainer) this.openContainer).isInUse()) {
+            SpongeImpl.getLogger().warn("This player is currently modifying an open container. This action will be delayed.");
+            Sponge.getScheduler().createTaskBuilder().delayTicks(0).execute(() -> {
+                this.closeInventory(); // Cause close event first. So cursor item is not lost.
```

Comment:
Should we perform any cause stack modifications here since this is performed a tick later, and more than likely the task would be redirected as sponge causing the closure and opening of the inventory?